### ISSUE TITLE
`entity_linker` pipeline is failing on ValueError 

### ISSUE THREAD

**Issue by Joozty** [explosion/spaCy#7065]
## How to reproduce the behaviour
When I try to train my EL pipeline I get an error that sentence is not in the list:

```python
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/path/venv/lib/python3.8/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/path/venv/lib/python3.8/site-packages/spacy/cli/_util.py", line 68, in setup_cli
    command(prog_name=COMMAND)
  File "/path/venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/path/venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/path/venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/path/venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/path/venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/path/venv/lib/python3.8/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/path/venv/lib/python3.8/site-packages/spacy/cli/train.py", line 59, in train_cli
    train(nlp, output_path, use_gpu=use_gpu, stdout=sys.stdout, stderr=sys.stderr)
  File "/path/venv/lib/python3.8/site-packages/spacy/training/loop.py", line 114, in train
    raise e
  File "/path/venv/lib/python3.8/site-packages/spacy/training/loop.py", line 98, in train
    for batch, info, is_best_checkpoint in training_step_iterator:
  File "/path/venv/lib/python3.8/site-packages/spacy/training/loop.py", line 194, in train_while_improving
    nlp.update(
  File "/path/venv/lib/python3.8/site-packages/spacy/language.py", line 1106, in update
    proc.update(examples, sgd=None, losses=losses, **component_cfg[name])
  File "/path/venv/lib/python3.8/site-packages/spacy/pipeline/entity_linker.py", line 229, in update
    sent_index = sentences.index(ent.sent)
ValueError:

Thomas Schippers introduced Kathleen Battle to his fellow conductor  James Levine  who selected Battle to sing in  Mahler 's  Symphony No. 8  at the  Cincinnati Symphony Orchestra 's M
ay Festival in 1974. is not in list

```

Based on the traceback I was able to find out [the code](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L221-L232) that causes it and here is a minimal working example:

```python
text = """
Thomas Schippers introduced Kathleen Battle to his fellow conductor  James Levine  who selected Battle to sing in  Mahler 's  Symphony No. 8  at the  Cincinnati Symphony Orchestra 's M
ay Festival in 1974.
"""
nlp = spacy.load("en_core_web_md", exclude=["tok2vec", "tagger", "parser", "tagger", "attribute_ruler", "lemmatizer", "ner"])

nlp.add_pipe("sentencizer")
ruler = nlp.add_pipe("entity_ruler",)
patterns = [{"label": "THING", "pattern": [{"LOWER": "symphony"}, {"LOWER": "no"}, {"LOWER": "."}, {"LOWER": "8"}]}]
ruler.add_patterns(patterns)

doc = nlp(text)
sentences = [s for s in doc.sents]
for ent in doc.ents:
    sent_index = sentences.index(ent.sent)
```

The problem is that sentence contains `No.` abbreviation and `doc.sents` recognized the whole text as one sentence unlike `ent.sent` that thinks it's two sentences.

## Your Environment

spaCy version    3.0.3
Location         /path/venv/lib/python3.8/site-packages/spacy
Platform         Linux-4.4.0-19041-Microsoft-x86_64-with-glibc2.29
Python version   3.8.5
Pipelines        en_core_web_md (3.0.0), en_core_web_sm (3.0.0)

**Comment by svlandeg** [comment]
This is an interesting edge case. Usually, entities probably shouldn't span multiple sentences, but that's exactly what happens here because "No. 8" is tagged as an entity, but at the same time the `sentencizer` defines a sentence split at the `.`.

`doc.sents` contains 3 sentences in your example text: 0-27, 27-43 and 43-44.
The "No. 8" entity is at indices 24-28, which means that `ent.sent` will actually return 2 sentences instead of 1: it will return the text 0-43. 

This behaviour is confirmed when looking at the [implementation](https://github.com/explosion/spaCy/blob/master/spacy/tokens/span.pyx#L359) of `Span.sent`: it takes the start offset for granted, and then starts looking from the `end` offset onwards to the first end of the sentence - effectively ignoring that there might be a sentence stop in the middle of the entity.

That said, I'm not sure what the best behaviour should be here. It kind of makes sense to return both sentences, but it feels like that also breaks the API contract somewhat, as you'd expect `ent.sent` to return exactly one sentence.

So either we somehow fix the behaviour in `Span`, or we'll need to address the (wrong) assumption that `ent.sent` always returns exactly one sentence - because the latter assumption is what the EL code relies on.

Curious to hear what others think!

**Comment by honnibal** [comment]
There’s no perfect solution here. Even if we constrain the NER to prevent it from crossing sentence boundaries, you can always set the sentences second and end up with the same inconsistency.

Here's a full list of options as I see them:

1. Raise an error
2. Return an empty span
3. Return a span with just the entity
4. Return the first sentence.
5. Return all sentences including words of the entity.
6. Adjust the sentence boundary so that it includes the entity.

I think 4 and 6 are the best options. We plainly can’t satisfy both constraints that the sentence should contain the entity and that the “sentence” should be a real sentence (as in `doc.sents`). I suggest we do 4, which I believe is the current behaviour? We should then document this clearly, and suggest the following fix if you need to ensure the sentence includes the full entity:

```
sent = ent.sent
sent = doc[sent.start : max(sent.end, ent.end)]
```

**Comment by svlandeg** [comment]
Ok. The current behaviour is 5 btw - it searches backwards from the start of the entity to the beginning of a sentence, and forward from the _end_ of the entity to the end of a sentence, ignoring the sentence stop in the middle and effectively returning the two sentences that have part of the entity in it.

I'll make a quick PR & adjust the docs.

**Comment by honnibal** [comment]
@svlandeg Great, thanks! I hadn't realised that the current behaviour was 5. I'll leave this open a little longer so we can gather objections, but your fix looks good.

**Comment by Joozty** [comment]
I have kind of different example which fails now (tried code from master) but it's working in 3.0.3. Is it intended because I kind of got lost in what agreed upon here


```python
import spacy

nlp = spacy.load("en_core_web_md", exclude=["tok2vec", "tagger", "parser", "tagger", "attribute_ruler", "lemmatizer", "ner"])

nlp.add_pipe("sentencizer")
ruler = nlp.add_pipe("entity_ruler",)
patterns = [{"label": "THING", "pattern": [{"TEXT": "("}, {"TEXT": "r"}, {"TEXT": ")"}]}]
ruler.add_patterns(patterns)

doc = nlp("""32nd  terrebonne  joseph l. waitz, jr. (r) 33rd  allen  h. todd nesom (d) 34th  st. bernard  perry m. nicosia (d) 35th  grant""")

sent_ents = [e for s in doc.sents for e in s.ents]
doc_ents = [ent for ent in doc.ents]

if len(sent_ents) != len(doc_ents):
    print('len dont match')
```

**Comment by svlandeg** [comment]
Hm, yes, good point @Joozty , I think this is an unfortunate side effect of having an entity that crosses sentence boundaries: neither of the sentences will consider this entity as being "in" that particular sentence.

I'm not sure there's a way to resolve that though...

**Comment by Joozty** [comment]
I thought so. There is no simple solution for this I guess and it's an edge case. 

I was just trying to train my EL model and run into this [error](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L391). Which was kind of hard to debug for me. From the user point of view, I would expect that once my Example object has entities EL pipeline will not fail. (you may / may not agree with me). What if these entities were skipped (empty kb_id or NIL) in evaluation?

Something like this:


```python
  count_ents = len([ent for doc in docs for ent in doc.ents])

  
  # btw not sure if this can happen
  if count_ents < len(kb_ids):
      raise ValueError(Errors.E148.format(ents=count_ents, ids=len(kb_ids)))
  i = 0
  for doc in docs:
      for ent in doc.ents:
          if ent.end_char > ent.sent.end_char:
			# silently ignore this edge case or warn user without failing
			continue
		  kb_id = kb_ids[i]
          i += 1
          for token in ent:
              token.ent_kb_id_ = kb_id
```

btw. why the sentences in the text are split the way they are? `(` is not char en

**Comment by svlandeg** [comment]
Ah, hm, yea ideally you wouldn't run into error `E148` because of this. The EL code was written with the assumption in mind that an entity is always part of a sentence. I'll reopen this issue and put it on my TODO list to make sure I revise that.

**Comment by svlandeg** [comment]
Hi @Joozty : the fix will be available from 3.0.6 onwards!
